lib/core/src/Cardano/Wallet/Primitive/AddressDiscovery/Delegation.hs

```diff
@@ -0,0 +1,464 @@
+
+        findOut ix = case map mkPointer pointerOuts of
+            (x:_) -> x
+            _     -> error $ mconcat
+                [ "couldn't find pointer output for ix "
+                , show ix
+                , " with state "
+                , show $ state ds
+                ]
+          where
+            isOurOut (TxOut addr _b) =
+                case (paymentKeyFingerprint @k $ keyAtIx ds ix, paymentKeyFingerprint addr) of
+                (Right fp, Right fp')
+                    | fp == fp' -> True
+                    | otherwise -> False
+                _ -> False
+            mkPointer (txIx, TxOut _ tb) = PointerUTxO (TxIn h txIx) (TB.getCoin tb)
+            pointerOuts = filter (isOurOut . snd) $ zip [0..] outs
+
+    modifyState
+        :: (State -> State)
+        -> DelegationState k
+        -> DelegationState k
+    modifyState f s = s { state = f (state s) }
+
+    -- | Modifies the "isKey0Reg" of the `More` constructor.
+    modifyKey0 c s@(More i p _) = case c of
+        Delegate k
+            | k == acct 0 -> More i p True
+            | otherwise   -> s
+        DeRegisterKey k
+            | k == acct 0 -> More i p False
+            | otherwise   -> s
+        _                 -> s
+      where
+        acct = toRewardAccount . keyAtIx s0 . toEnum
+    modifyKey0 _ s = s
+
+    lastActiveKey s' = toRewardAccount . keyAtIx s' <$> lastActiveIx s'
+    nextKey s' = toRewardAccount . keyAtIx s' $ nextKeyIx s'
+
+--------------------------------------------------------------------------------
+-- Operations
+--------------------------------------------------------------------------------
+
+-- | All stake keys worth listing to the user.
+--
+-- May include:
+-- 1. Active stake keys
+-- 2. The next un-active key
+--
+-- NOTE: In theory we might want also present stake keys that are unexpectedly
+-- registered, as they could be de-registered to reclaim the deposit, but this
+-- should in-practice never happen.
+--
+-- If @sn@ denotes the state with @n@ registered and delegating keys:
+-- >>> presentableKeys s0
+-- [0]
+-- >>> presentableKeys s1
+-- [0, 1]
+-- >>> presentableKeys s2
+-- [0, 1, 2]
+presentableKeys :: SoftDerivation k => DelegationState k -> [k 'AddressK XPub]
+presentableKeys s = case lastActiveIx s of
+    Just i -> map (keyAtIx s) [minBound .. (succ i)]
+    Nothing -> [keyAtIx s minBound]
+
+-- Keys meant to be used in addresses.
+--
+-- If @sn@ denotes the state with @n@ registered and delegating keys:
+-- >>> usableKeys s0
+-- [0]
+-- >>> usableKeys s1
+-- [0]
+-- >>> usableKeys s2
+-- [0, 1]
+--
+-- Note that for @s0@, we have no active stake keys, but we still want to use
+-- key 0 as part of addresses.
+usableKeys :: SoftDerivation k => DelegationState k -> [k 'AddressK XPub]
+usableKeys s = case lastActiveIx s of
+    Just i -> map (keyAtIx s) [minBound .. i]
+    Nothing -> [keyAtIx s minBound]
+
+-- | For testing. Returns all registered and delegating stake keys.
```

Comment:
Related to the topic of how key 0 should be handled:

This doc-comment is technically a lie. Key 0 might not be delegating.

1. This is misleading
2. I do hope the tests still pass if this might skip key 0 accordingly

Reply:
Added comments both here and for `usableKeys`, and the tests still pass if `activeKeys` skips key 0 accordingly.